contrib/apache-httpd/mod_arpa2_tlspool.c

```diff
@@ -0,0 +1,314 @@
+    if (strcEQ(s->server_hostname, tlsdata->localid)) {
+        tlspool_server_config *pConfig = our_sconfig(s);
+        ssl_verify_t mode = pConfig->nVerifyClient;
+        // unset previous set flags;
+        tlsdata->flags &= ~(PIOF_STARTTLS_REQUEST_REMOTEID | PIOF_STARTTLS_IGNORE_REMOTEID);
+        switch (mode) {
+            case SSL_CVERIFY_REQUIRE:
+            // default on tlspool_starttls
+                break;
+            case SSL_CVERIFY_OPTIONAL:
+                tlsdata->flags |= PIOF_STARTTLS_REQUEST_REMOTEID;
+                break;
+            default:
+                tlsdata->flags |= PIOF_STARTTLS_IGNORE_REMOTEID;
+                break;
+        }
+        return 1;
+    }
+    return 0;
+}
+
+static char tmp[4096];
+
+typedef struct {
+    int plainfd;
+    conn_rec *conn;
+} privdata_t;
+
+static int namedconnect_vhost (starttls_t *tlsdata, void *p)
+{
+    privdata_t* privdata = (privdata_t *) p;
+strcpy(tmp, tlsdata->localid);
+    if (tlsdata->localid[0] != '\0') {
+        ap_vhost_iterate_given_conn(privdata->conn, func_cb, tlsdata);
+    }
+
+#if !defined(WINDOWS_PORT)
+	int soxx[2];
+	if (socketpair (AF_UNIX, SOCK_STREAM, 0, soxx) == 0)
+#else /* WINDOWS_PORT */
+	// https://github.com/ncm/selectable-socketpair
+	extern int dumb_socketpair(SOCKET socks[2], int make_overlapped);
+	SOCKET soxx[2];
+	if (dumb_socketpair(soxx, 1) == 0)
+#endif /* WINDOWS_PORT */
+	{
+		privdata->plainfd = soxx [1];
+		return soxx [0];
+	}
+}
+
+/*
+ * This routine is called just after the server accepts the connection,
+ * but before it is handed off to a protocol module to be served.  The point
+ * of this hook is to allow modules an opportunity to modify the connection
+ * as soon as possible. The core server uses this phase to setup the
+ * connection record based on the type of connection that is being used.
+ *
+ * This is a RUN_ALL hook.
+ */
+static int tlspool_pre_connection(conn_rec *c, void *csd)
+{
+    char *note;
+
+    if (pMainConfig->bEnabled) {
+        apr_socket_t *apr_socket = (apr_socket_t *) csd;
+        apr_os_sock_t os_sock;
+        apr_os_sock_get(&os_sock, apr_socket);
+        int cnx = (int) os_sock;
+        privdata_t privdata = { -1, c };
+
+        tlsdata_now = tlsdata_srv;
+        switch (pMainConfig->nVerifyClient) {
+            case SSL_CVERIFY_REQUIRE:
+                // default on tlspool_starttls
+                break;
+            case SSL_CVERIFY_OPTIONAL:
+                tlsdata_now.flags |= PIOF_STARTTLS_REQUEST_REMOTEID;
+                break;
+            default:
+                tlsdata_now.flags |= PIOF_STARTTLS_IGNORE_REMOTEID;
+                break;
+        }
+        note = apr_psprintf(c->pool, "tlspool_pre_connection: c = %pp, pool = %pp, old = %d, nVerifyCLient = %d, flags = 0x%08x",
+                        (void*) c, (void*) c->pool, cnx, pMainConfig->nVerifyClient, tlsdata_now.flags);
+        trace_nocontext(c->pool, __FILE__, __LINE__, note);
```

Comment:
debugging output, right?  does it always need to be output?  or does Apache filter it as the debug logging level?

Reply:
(probably applies elsewhere too)